File: LeetCode/Problems/Python/127. Word Ladder.py
#BFS
#Time: O(M^2*N)
#Space: O(M^2*N)
class Solution:
    def ladderLength(self, beginWord: str, endWord: str, wordList: List[str]) -> int:
        if not beginWord or not endWord or not wordList or endWord not in wordList:
            return 0
        
        graph = defaultdict(set)
        
        #for loop M*N, create new string M => O(M^2*N)
        for w in wordList:
            for i in range(len(w)):
                graph[w[:i]+"*"+w[i+1:]].add(w)
   

        queue = deque()
        queue.append([beginWord,1])
        visited = set([beginWord])
        
        #worst case also O(M^2*N)
        while queue:
            word, steps = queue.popleft()
            for i in range(len(word)):
                for node in graph[word[:i]+"*"+word[i+1:]]:
                    if node not in visited:
                        visited.add(node)
                        if node == endWord:
                            return steps+1
                        queue.append([node, steps+1])
                 
        return 0
        

File: LeetCode/Problems/C++/127. Word Ladder.cpp
class Solution {
public:
    int ladderLength(string beginWord, string endWord, vector<string>& wordList) {
        if(!beginWord.size() || !endWord.size() || !wordList.size()) return 0;
        if(std::find(wordList.begin(), wordList.end(), endWord) == wordList.end()) return 0;
        
        unordered_map<string, vector<string>> graph;
        
        for(int i=0;i<wordList.size();i++){
            for(int j=0;j<beginWord.size();j++){
                graph[wordList[i].substr(0,j) + "*" + wordList[i].substr(j+1)].push_back(wordList[i]);
            }
        }
        
        unordered_set<string> visited;
        visited.insert(beginWord);
        
        queue<pair<string, int>> bfs_queue;
        bfs_queue.push({beginWord, 1});
        
        while(!bfs_queue.empty()){
            string curr = bfs_queue.front().first;
            int steps = bfs_queue.front().second;
            bfs_queue.pop();
            
            for(int i=0;i<curr.size();i++){
                vector<string> neighbor = graph[curr.substr(0,i) + "*" + curr.substr(i+1)];
                for(auto it=neighbor.begin(); it!=neighbor.end();it++){
                    if(*it==endWord) return steps+1;
                    if(visited.find(*it)==visited.end()){
                        visited.insert(*it);
                        bfs_queue.push({*it, steps+1});
                    }     
                }
            }   
        }
        return 0;
    }
};

class Solution {
public:
    int ladderLength(string beginWord, string endWord, vector<string>& wordList) {
        if(!beginWord.size() || !endWord.size() || !wordList.size()) return 0;
        if(std::find(wordList.begin(), wordList.end(), endWord) == wordList.end()) return 0;
        
        unordered_map<string, vector<string*>> graph;
        
        for(int i=0;i<wordList.size();i++){
            for(int j=0;j<beginWord.size();j++){
                graph[wordList[i].substr(0,j) + "*" + wordList[i].substr(j+1)].push_back(&wordList[i]);
            }
        }
        
        unordered_set<string*> visited;
        visited.insert(&beginWord);
        
        queue<pair<string, int>> bfs_queue;
        bfs_queue.push({beginWord, 1});
        
        while(!bfs_queue.empty()){
            string curr = bfs_queue.front().first;
            int steps = bfs_queue.front().second;
            bfs_queue.pop();
            
            for(int i=0;i<curr.size();i++){
                vector<string*>* neighbor = &graph[curr.substr(0,i) + "*" + curr.substr(i+1)];
                for(auto it=(*neighbor).begin(); it!=(*neighbor).end();it++){
                    if(**it==endWord) return steps+1;
                    if(visited.find(*it)==visited.end()){
                        visited.insert(*it);
                        bfs_queue.push({**it, steps+1});
                    }     
                }
            }   
        }
        return 0;
    }
};